### PR TITLE
dashboard: Add support for firing events

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

Add support for firing events

We now fire the following events:

Intended to be consumed by a `websocket` version of the `devices` endpoint.  These events are for the longer term, stored state of the dashboard entry.
- EVENT_ENTRY_ADDED - A new dashboard entry has been added
- EVENT_ENTRY_REMOVED - A dashboard entry has been removed
- EVENT_ENTRY_UPDATED - A dashboard entry has been updated

Intended to be consumed by a `websocket` version of the `ping` endpoint. These events are for the short term transient state of the dashboard entry. This event may change in the future to include more than the online/offline state if we need to know more about what the device is doing but currently this is no need to extend it.
- EVENT_ENTRY_STATE_CHANGED - The state of a dashboard entry has changed (previously ping result)

Current nothing listens for these events yet, there are no events for discovery yet (aka importable)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/dashboard/web_server.py [hunk 1]
```diff
@@ -785,9 +796,6 @@ def post(self, configuration=None):
             if build_folder is not None:
                 shutil.rmtree(build_folder, os.path.join(trash_path, name))
 
-        # Remove the old ping result from the cache
-        DASHBOARD.ping_result.pop(configuration, None)
```
Comment (bdraco) [inline esphome/dashboard/web_server.py:789]: The object gets deleted now so no need to remove it here

File: esphome/dashboard/entries.py [hunk 2]
```diff
@@ -3,24 +3,78 @@
...
+
+def bool_to_entry_state(value: bool) -> EntryState:
+    """Convert a bool to an entry state."""
+    return _BOOL_TO_ENTRY_STATE[value]
+
+
+def entry_state_to_bool(value: EntryState) -> bool | None:
+    """Convert an entry state to a bool."""
+    return _ENTRY_STATE_TO_BOOL[value]
+
```
Comment (bdraco) [inline esphome/dashboard/entries.py:58]: The ping api expects null/true/false so we need to be able to convert

File: esphome/dashboard/status/mdns.py [hunk 3]
```diff
@@ -22,16 +24,16 @@ def __init__(self) -> None:
...
         self.host_mdns_state: dict[str, bool | None] = {}
-        # This is the hostnames to filenames mapping
-        self.host_name_to_filename: dict[str, str] = {}
-        self.filename_to_host_name: dict[str, str] = {}
+        # This is the hostnames to path mapping
+        self.host_name_to_path: dict[str, str] = {}
+        self.path_to_host_name: dict[str, str] = {}
         # This is a set of host names to track (i.e no_mdns = false)
         self.host_name_with_mdns_enabled: set[set] = set()
         self._loop = asyncio.get_running_loop()
```
Comment (bdraco) [inline esphome/dashboard/status/mdns.py:32]: All these indices should be moved into `DashboardEntries` in a future PR so we don't duplicate work

File: esphome/dashboard/status/mdns.py [hunk 4]
```diff
@@ -22,16 +24,16 @@ def __init__(self) -> None:
...
 
-    def filename_to_host_name_thread_safe(self, filename: str) -> str | None:
-        """Resolve a filename to an address in a thread-safe manner."""
-        return self.filename_to_host_name.get(filename)
+    def get_path_to_host_name(self, path: str) -> str | None:
+        """Resolve a path to an address in a thread-safe manner."""
+        return self.path_to_host_name.get(path)
```
Comment (bdraco) [inline esphome/dashboard/status/mdns.py:36]: This should move to `DashboardEntries` as well in a future PR

**Review by balloob — APPROVED** [review]

**Pull request merged**